docs/design/log_tracing.md

```diff
@@ -0,0 +1,112 @@
+investigated would be `info`. In case a tracing system (like Jaeger) is being used, the information is filtered on a
+different level (the tracing system), and the default setting would become `info`.
+
+A default setting for developer workload would be `info`. Alternate settings are up to the developer.
+
+## Instrumentation
+
+Function calls can be wrapped with the `#[instrument]` attribute, which allows tracing calls to functions, including
+their arguments and return values.
+
+Adding instrumentation like this adds some overhead. So it should be applied with care. However, on the other side, it
+can be used to profile the application, also in production environments. This can be a considerable help, when things
+go wrong in production.
+
+### Arguments
+
+Arguments get serialized in a string format which can either be printed other sent to a tracing system. However, this
+also means that the data needs to be processed. The bigger the data, the bigger the overhead. Arguments can be skipped,
+and in most cases larger structs and `self` can and should be skipped:
+
+```rust
+impl Example {
+  #[instrument(skip(self, data))]
+  fn example(&self, data: LargeStruct, flag: bool) {
+    // ...
+  }
+}
+```
+
+In cases where partial information of bigger structs may be relevant, it can be extracted using `field`:
+
+```rust
+impl Example {
+  #[instrument(skip(items), field(num_items=items.len()))]
+  fn example(items: Vec<LargeStruct>) {
+    // ...
+  }
+}
+```
+
+### Return values and errors
+
+The `instrument` macro allows automatically extracting return values and errors. Return values get logged with the same
+level the function gets logged, and errors as `error`. This can be overridden but is a reasonable default. For `err` it
+is required that the function returns a `Result`.
+
+Contrary to the documentation, using `ret` logs both `Result::Ok` and `Result::Err` (using the same level). Adding `err`
+logs the error differently (level and format). Also see: https://github.com/tokio-rs/tracing/issues/2963.
+
+```rust
+#[instrument(ret)]
+fn example() -> Result<bool, Error> {}
+```
+
+As the idea is to keep `error` level messages for things that potentially disrupt the normal operation of the
+application, just using `ret` is fine. Unless, of course, returning an error would actually cause a potential disruptive
+situation.
+
+This also means that the result will be logged or sent to the tracing system, so it should not be used for larger
+structs. In cases like that, use `err`, but lower the level to `info`:
+
+```rust
+#[instrument(err(level=tracing::Level::INFO))]
+fn example() -> Result<LargeStruct, Error> {}
+```
+
+## println and panic
+
+An application should not panic during normal runtime. Also, all output should use logging or tracing.
+
+There is one exception to this rule (aside from tests): the start of the application. During the logging/tracing system
```

Comment:
what about migrations (I see usage of .panic there)

Reply:
Good point. I expanded that section a bit.